File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e6))


def find(num):
    if numbers[num] != num:
        numbers[num] = find(numbers[num])
    return numbers[num]


def union(n1, n2):
    n1, n2 = find(n1), find(n2)

    if n1 == n2: return
    elif n1 < n2: numbers[n2] = n1
    else: numbers[n1] = n2


n, m = map(int, input().split())
numbers = [i for i in range(n + 1)]

for i in range(m):
    ipt = list(map(int, input().split()))

    if ipt[0] == 0:
        union(ipt[1], ipt[2])
    else:
        print('YES' if find(ipt[1]) == find(ipt[2]) else 'NO')

File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
#include <vector>

using namespace std;

vector<int> parent;

int find(int p) {
  if (parent[p] != p) parent[p] = find(parent[p]);
  return parent[p];
}

void _union(int p, int q) {
  p = find(p);
  q = find(q);
  if (p == q) return;
  if (p < q)
    parent[q] = p;
  else
    parent[p] = q;
  return;
}

int main() {
  int n, m;
  scanf("%d %d", &n, &m);

  for (int i = 0; i < n + 1; i++) {
    parent.push_back(i);
  }

  for (int i = 0; i < m; i++) {
    int command, a, b;
    scanf("%d %d %d", &command, &a, &b);

    switch (command) {
      case 0:
        _union(a, b);
        break;
      case 1:
        if (find(a) == find(b))
          printf("YES\n");
        else
          printf("NO\n");
        break;
    }
  }

  return 0;
}
